# Zero to Hero

## 1.06 Hackeando Kaggle

Voy a **hackear** Kaggle  para descubrir cuantos  "BAJA+2"  hay en el mes 202101, que es donde  *no* tengo la clase

La idea es subir a Kaggle una prediccion que tenga TODOS 1,  que se le envíe estímulo a todos
<br>
Es obvio que esa prediccion va a dar una pésima ganancia
<br>
Pero me va a permitir tener un sistema de dos ecuaciones con dos incognitas, y despejando tendre la cantidad de BAJA+2 que hay en los datos del futuro

In [3]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria  rpart
library( "rpart.plot")

options(repr.plot.width=20, repr.plot.height=10) 
setwd("C:/Users/idott/LI")  #Aqui se debe poner la ruta de la PC local

used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  647835 34.6 1318414    70.5 1318414  70.5
Vcells 1152697  8.8 8388608    64.0 1707891  13.1

Leo el dataset

In [4]:
dataset <- fread("./datasets/dataset_pequeno.csv")   #cargo el dataset

Cuento la clase

In [5]:
dataset[  , .N, list( foto_mes, clase_ternaria) ]

foto_mes clase_ternaria N     
1 202107   CONTINUA       162336
2 202107   BAJA+1           1078
3 202107   BAJA+2           1268
4 202109                  165237

Defino  el data.table  dfuturo que va a tener solo los datos de  202109

In [6]:
dfuturo <-  dataset[ foto_mes==202109 ]

In [7]:
dfuturo

numero_de_cliente foto_mes active_quarter cliente_vip internet
1      31116053          202109   1              0           0       
2      31116803          202109   1              0           0       
3      31117730          202109   1              0           0       
4      31117908          202109   1              0           0       
5      31117977          202109   1              0           0       
6      31120119          202109   1              0           0       
7      31121017          202109   1              0           0       
8      31121264          202109   1              0           0       
9      31121718          202109   1              0           0       
10     31122951          202109   1              0           0       
11     31123702          202109   1              0           0       
12     31123830          202109   1              0           0       
13     31124501          202109   1              0           0       
14     31124827          202109   1              0           0       
15     31125498          202109   1              0           0       
16     31125518          202109   1              0           0       
17     31126001          202109   1              0           0       
18     31126199          202109   1              0           0       
19     31126633          202109   1              0           0       
20     31127136          202109   1              0           0       
21     31127195          202109   1              0           0       
22     31129377          202109   1              0           0       
23     31130581          202109   1              0           0       
24     31131104          202109   1              0           0       
25     31132831          202109   1              0           0       
26     31132989          202109   1              0           0       
27     31133028          202109   1              0           0       
28     31133769          202109   1              0           0       
29     31133788          202109   1              0           0       
30     31133818          202109   1              0           0       
⋮      ⋮                 ⋮        ⋮              ⋮           ⋮       
165208 214662165         202109   0              0           0       
165209 214663043         202109   1              0           0       
165210 214663340         202109   1              0           0       
165211 214663428         202109   1              0           0       
165212 214665787         202109   0              0           0       
165213 214666202         202109   1              0           0       
165214 214677423         202109   0              0           0       
165215 214687856         202109   1              0           0       
165216 214694024         202109   1              0           0       
165217 214707703         202109   1              0           0       
165218 214710546         202109   0              0           0       
165219 214721787         202109   1              0           0       
165220 214724936         202109   1              0           0       
165221 214729140         202109   1              0           0       
165222 214729673         202109   0              0           0       
165223 214738852         202109   0              0           0       
165224 214740263         202109   0              0           0       
165225 214745168         202109   1              0           0       
165226 214749857         202109   0              0           0       
165227 214758966         202109   1              0           0       
165228 214767247         202109   1              0           0       
165229 214774866         202109   1              0           0       
165230 214774985         202109   0              0           0       
165231 214782860         202109   1              0           0       
165232 214787203         202109   0              0           0       
165233 214798593 

cuento la cantidad de lineas del dataset

In [8]:
nrow( dfuturo )

[1] 165237

Creo un vector que sean los  ids

In [9]:
vector_ids  <-   dfuturo[ , numero_de_cliente]

In [10]:
head( vector_ids)

[1] 31116053 31116803 31117730 31117908 31117977 31120119

Creo un vector de todos unos, con la cantidad de registros que tiene dfuturo
<br>
utilizo la instruccion  **rep()**  de R  https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/rep

In [11]:
vector_enviar <-  rep( 1,  nrow( dfuturo))

In [12]:
head( vector_enviar )

[1] 1 1 1 1 1 1

In [13]:
length( vector_enviar)

[1] 165237

Excelente, tiene tantos unos como registros tiene dfuturo

creo una tabla con  lo que voy a enviar a Kaggle

In [14]:
tabla_final  <-   as.data.table(  list(  "numero_de_cliente"= vector_ids,
                                         "Predicted"=         vector_enviar))

In [15]:
head( tabla_final)

numero_de_cliente Predicted
1 31116053          1        
2 31116803          1        
3 31117730          1        
4 31117908          1        
5 31117977          1        
6 31120119          1

In [16]:
#genero el archivo para Kaggle
#creo la carpeta donde va el experimento
dir.create( "./exp/",  showWarnings = FALSE ) 
dir.create( "./exp/ZH2016/", showWarnings = FALSE )

#genero el archivo para Kaggle
fwrite(tabla_final, 
        file= "./exp/ZH2016/todos_unos.csv",
        sep=  "," )

Subo el archivo a Kaggle y me fijo cuanta ganancia me da en el Public Leaderboard

da una ganancia de **-368.11331**   ( menos 368 millones )

### el hackeo

Ahora pasamos a calcular la cantidad de BAJA+2  que hay en los datos de septiembre-2021,  hackeamos Kaggle

Tengo dos incognitas, por un lado lo que llamo POS , los positivos, las "BAJA+2"
<br>
Por otro lado, lo que llamo NEG, los negativos,  los "BAJA+1" y "CONTINUA"

Se que se cumplen estas dos cosas:
* POS + NEG = 165237
* 117000\*POS - 3000\*NEG =-368113310

Es un sistema de dos ecuaciones con dos incógnitas, has multiples formas de resolverlo, la primera sería despejar POS.
<br>
pero utilizamos esta página  https://keisan.casio.com/exec/system/14013211335541

Al correrlo obtenemos que  POS = 1063, o sea **las BAJA+2  de 202109  son  1063**

¿Qué nos dicen estas 1063 BAJA+2 de septiembre del 2021? ¿Explican algo?
<br>
Las BAJA+2 que hay en julio del 2021 son  1268
<br>
1268 > 1063
<br>
Con lo cual es de esperar mayores ganancias de los modelos predictivos en 202107  que en 202109,  va a irnos PEOR en el Public Leaderboard que en nuestros traning/testing.